src/sage/knots/free_knotinfo_monoid.py

```diff
@@ -0,0 +1,479 @@
+                if c > former_cr and res:
+                    k = self._check_elements(knot, tuple(res))
+                    if k:
+                        # matching item found
+                        return tuple([k])
+                    former_cr = c
+
+            return tuple(sorted(set(res)))
+
+        hp = knot.homfly_polynomial(normalization='vz')
+        return search_composition(13, hp)
+
+    def from_knot(self, knot, unique=True):
+        """
+        Create an element of this abelian monoid from ``knot``.
+
+        INPUT:
+
+        - ``knot`` -- an instance of :class:`~sage.knots.knot.Knot`
+
+        - ``unique`` -- boolean (default is ``True``). This only affects the case
+          where a unique identification is not possible. If set to ``False`` you
+          can obtain a matching list (see explanation of the output below)
+
+        OUTPUT:
+
+        An instance of the element class of ``self`` per default. If the keyword
+        argument ``unique`` then a list of such instances is returned.
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: K = KnotInfo.K5_1.link().mirror_image()
+            sage: FKIM.from_knot(K)
+            KnotInfo['K5_1m']
+
+            sage: # optional - database_knotinfo
+            sage: K = Knot(KnotInfo.K9_12.braid())
+            sage: FKIM.from_knot(K)                   # long time
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: this (possibly non prime) knot cannot be identified uniquely by KnotInfo
+            use keyword argument `unique` to obtain more details
+            sage: FKIM.from_knot(K, unique=False)     # long time
+            [KnotInfo['K4_1']*KnotInfo['K5_2'], KnotInfo['K9_12']]
+        """
+        hp = knot.homfly_polynomial(normalization='vz')
+        num_summands = sum(e for f, e in hp.factor())
+        if num_summands == 1:
+            return knot.get_knotinfo()
+
+        res = self._from_knot(knot)
+        if res:
+            if len(res) == 1:
+                if unique:
+                    return res[0]
+                else:
+                    return [res[0]]  # to be consistent with get_knotinfo
+            k = self._check_elements(knot, res)
+            if k:
+                if unique:
+                    return k
+                else:
+                    return[k]  # to be consistent with get_knotinfo
+
+        if res and not unique:
+            return sorted(list(set(res)))
+        if unique and len(res) > 1:
+            non_unique_hint = '\nuse keyword argument `unique` to obtain more details'
+            raise NotImplementedError('this (possibly non prime) knot cannot be identified uniquely by KnotInfo%s' % non_unique_hint)
+        else:
+            raise NotImplementedError('this (possibly non prime) knot cannot be identified by KnotInfo')
```

Comment:
```suggestion
        raise NotImplementedError('this (possibly non prime) knot cannot be identified by KnotInfo')
```
Personal preference, but I don't like such `else:` statements to end as it could be that one of the `if` statements (in the future) accidentally falls through resulting in the method returning `None`.